### ISSUE TITLE
mp3 file length displaying incorrectly

### ISSUE THREAD

**Issue by enwardy** [paradisec-archive/nabu#615]
GS1-H10b-1

the wav length is displayed as 2:24, the mp3 as 47:04. ..But then the wav being 1.5GB, I went in and played it, and it seems to be the full length of 47mins. (I haven't actually downloaded it to check the exact length, but based on playing it in the preview it is a lot lot longer than 2:24).

**Comment by enwardy** [comment]
another example (and it seems like there are quite a few in this collection): GS1-H2b

**Comment by JonoGillett** [comment]
I can confirm that the WAV files are reporting an incorrect duration at a filesystem level, which indicates that the files may have invalid or corrupt header information.

```
deploy@paradisec:/srv/catalog/GS1/H10b$ ffprobe -v 1 GS1-H10b-1.mp3 -show_format_entry duration
duration=2824.124082 (approx 47 mins)
deploy@paradisec:/srv/catalog/GS1/H10b$ ffprobe -v 1 GS1-H10b-1.wav -show_format_entry duration
duration=144.270490 (approx 2 mins)
```
and similarly
```
deploy@paradisec:/srv/catalog/GS1/H2b$ ffprobe -v 1 GS1-H2b-1.wav -show_format_entry duration
duration=0.790677
deploy@paradisec:/srv/catalog/GS1/H2b$ ffprobe -v 1 GS1-H2b-1.mp3 -show_format_entry duration
duration=2494.119184
```

**Comment by enwardy** [comment]
@JonoGillett , further to these wavs displaying incorrect file duration, I have found a large amount of mp3s which also seem to be displaying incorrect file duration. 

The file duration as displayed by Nabu (and as displayed and played in the Nabu file-player) anything from a fraction  of a second shorter than the file actually is, to around 20 seconds (LD1-md0202-A.mp3, which Nabu displays as 02:22:21.709, but when downloaded and played it is actually 02:22:41...., the same duration as the wav)

The amount of inaccurate duration reported (mp3 reported by Nabu as shorter than it actually is) is in exact correlation to the file length. Longer file = more inaccurately under-reported mp3 duration. This is represented in this spreadsheet + chart:
[mp3s_shorter.xlsx](https://github.com/nabu-catalog/nabu/files/1140944/mp3s_shorter.xlsx)

**Comment by j03w** [comment]
looks like a bug in ffmpeg upgrading the version should fix this issue https://trac.ffmpeg.org/ticket/3428

**Comment by enwardy** [comment]
@JonoGillett , so is ffmpeg version upgraded now? 
You are doing that (or have done it, and if so as of when, so I can check the reports history for a resolution of the error), or we need to do it? 
where is the software running from, to produce the report?

**Comment by JonoGillett** [comment]
@enwardy @j03w The issue linked above is out of date by about 3 years. We are using a version of ffmpeg from earlier this year. I can see if there is another newer version, but I doubt it would fix the problem.

**Comment by JonoGillett** [comment]
Strangely, running ffprobe on the original files mentioned in this issue, they now both report approximately the same - 2m28s - which is corroborated by listening to the file in VLC media player.

It seems they were updated around the 15th of June.

```
deploy@paradisec:/srv/catalog/GS1/H10b$ ffprobe -v 1 GS1-H10b-1.mp3 -show_format_entry duration
duration=148.741224 (~ 2m 28s)
deploy@paradisec:/srv/catalog/GS1/H10b$ ffprobe -v 1 GS1-H10b-1.wav -show_format_entry duration
duration=148.716583
```

and likewise for the H2b files (updated June 2nd):

```
deploy@paradisec:/srv/catalog/GS1/H2b$ ffprobe -v 1 GS1-H2b-1.mp3 -show_format_entry duration
duration=259.892245 (~ 4m 20s)
deploy@paradisec:/srv/catalog/GS1/H2b$ ffprobe -v 1 GS1-H2b-1.wav -show_format_entry duration
duration=259.874135
```

**Comment by enwardy** [comment]
yep, well many of the truncated wavs in this issue were occurring over the last several years, not necessarily recently. So the ffprobe issue of 3 years ago might still have been part of the issue. I'm trying to look further into it. Do you have any record of when you updated the versions of ffprobe over the last 3 years?

**Comment by JonoGillett** [comment]
We've updated it a few times recently - mostly in the hopes of addressing this issue. My memory/records are pretty vague before about December 2016.
(note: I'm not 100% sure of the exact versions, but we would have upgraded to the latest available version each time we updated, so based on when each update happened the below should be correct)

- at some point in 2015/early 2016 it was updated to "N-76944-g15206ff" (released in late 2015)
- in March 2017 we updated it from "N-76944-g15206ff" to 3.1.7 (released in Feb 2017)
- in October 2017 we updated it from 3.1.7 to 3.1.11 (released in Sep 2017)

**Comment by j03w** [comment]
Apologies for my last comment I didn't know what version of ffprobe we are using and just assumed that it's old version.

Either way after more investigation I'd like to conclude that this is not an issue on nabu but the system that actually process the audio files. The wave files actually got cut to the length reported through nabu.

If there's nothing more we can do I might close this off.

@enwardy  @JonoGillett

**Comment by enwardy** [comment]
I need to check if this is still an issue..

**Comment by enwardy** [comment]
Still an issue. I will post some example files here later in the week

**Comment by JonoGillett** [comment]
@enwardy what is the current action expected from us on this issue?

Do you want us to investigate alternatives to ffprobe?
Or do a re-calculate of the essence metadara for any files that are incorrect?
Or something else?

**Comment by enwardy** [comment]
@JonoGillett we've discussed this here and think, yes, perhaps a good place to start would be a re-calculate of the essence metadata for a small handful of mp3 files that are displaying incorrectly, and we'll see if it fixes it. See list below of files; start with those and we'll go from there.

Incidentally, the dates of many of these files look to have been last modified around middle of 2016 (the ones we looked at specifically were between March-July 2016), which we notice was a period when we were having multiple other issues occurring with the repository, some at least related to work Andrew Grimm was doing at the time.

CLV1-151-D.mp3
LD1-md0202-A.mp3
JM1-21-A.mp3
DG3-018-A.mp3

**Comment by enwardy** [comment]
I've added the April 2018 milestone, and therefore inclusion in the planned upcoming sprint, with reference to the initial step, in my previous comment. Once that step is complete, we can assess what needs to be done with this issue overall

**Comment by enwardy** [comment]
I've just checked, and yes these still need investigating (see my comment above on 27 Feb for details/instructions).

Incidentally, I noticed whilst checking the above listed files to test, that the mp3 duration appears incorrectly not only in Nabu, but also in mac finder & mac file info. However upon opening the file in quicktime or iTunes, the full and correct duration is evident.

**Comment by JonoGillett** [comment]
@enwardy just to confirm: you would like us to recalculate the metadata for some selection of files using the current methodology (i.e. ffprobe)? And then see if these now report the correct results?

I'm working under the assumption that the ideal targets for the recalculation are some of the 'most incorrect' in the CSV.

**Comment by JonoGillett** [comment]
I've re-run ffprobe across each of the essences mentioned above, and they all report the same durations as before.

As you pointed out, @enwardy, many applications (not just ffprobe/ffmpeg) report the same 'wrong' duration. I tried the following command line tools: ffprobe, mp3info, mp3check, and soxi, and the only one which reported the 'correct' time was soxi (however all the rest are accepted standard tools as well).

Given that many applications (including Chrome's audio player and Windows Media Player) actually report the same value as their duration, I'm not sure whether this is a bug or whether there is just diversity across the MP3 ecosystem.

The standard MP3 duration calculation is to use CBR (constant bit rate) approximation. This is calculated as follows:
`(file size in kb / bit rate) = time (in seconds)`

so for CLV1-151-D.mp3 which is 48896714 kB (or 391173712 kbits) this would be:
`391173712 / 128k = 391173712 / 128000 = 3056.044625 seconds or ~50:56`

This is the value reported by Chrome's audio player, Windows Media Player and a few other apps. Even VLC media player initially shows this value, but then seems to re-calculate it (strangely to be 51:03 or 1 second longer than the WAV) before it begins playing.

We could change the tool which is used to inspect the file metadata, however then someone playing the audio in Chrome or WMP would see a shorter duration than listed in the metadata.

My guess is that one or both of these is true:
1. the duration reported is not exact, regardless of source. in which case this is an expected situation with the MP3 format
1. there is something wrong with the files which causes them to report differently in some systems compared to others. in which case this is an issue with these files

Regardless, this is an issue outside of the scope of the Nabu application.

Given that we've already spent a considerable amount of time investigating this situation from the Nabu side, I would recommend that the investigation be moved to the parts of the system which handle audio file conversion as that is likely the place where any issues could be found, and fixes implemented.

**Comment by johnf** [comment]
I suspect this is now resolved in the new pipeline

**Comment by NickWardPDSC** [comment]
probably a different issue, but the description matches this existing (closed) issue, so putting it here:

this mp3 displays as 20:16, but is actually one 5:04 (matching the wav)

https://catalog.paradisec.org.au/collections/DD1/items/1113
DD1-1113-LR.mp3 | audio/mpeg | 4.64 MB | 00:20:16.390 | View
DD1-1113-LR.wav | audio/vnd.wav | 167 MB | 00:05:03.992 | View

**Comment by johnf** [comment]
Unfortunately we'll always see issues like this dut to the nature of ffmpeg/mediainfo and media files.

@NickWardPDSC Do you think this is important enough that we should have a button where you can request a metadata reprobe? That would use the current latest tools?
